File: backend/app/services/bitacora_service.py
"""
Servicio para gestión de bitácora de acciones del sistema.
Registra todas las acciones de usuarios para auditoría y trazabilidad.
"""
from typing import Optional, Dict, Any
from datetime import datetime, timedelta
from sqlalchemy.orm import Session
import json
import logging

from app.db.models.bitacora import T_Bitacora
from app.repositories.bitacora_repository import BitacoraRepository
from app.constants.tipos_accion import TiposAccion, DESCRIPCIONES_TIPOS_ACCION

logger = logging.getLogger(__name__)


class BitacoraService:
    """Servicio para manejo de bitácora de acciones del sistema"""
    
    def __init__(self):
        self.repo = BitacoraRepository()
    
    async def registrar(
        self,
        db: Session,
        usuario_id: str,
        tipo_accion_id: int,
        texto: str,
        expediente_id: Optional[int] = None,
        info_adicional: Optional[Dict[str, Any]] = None
    ) -> T_Bitacora:
        """
        Registra una acción en la bitácora.
        
        Args:
            db: Sesión de base de datos
            usuario_id: ID del usuario que realiza la acción (cédula)
            tipo_accion_id: ID del tipo de acción (usar constantes de TiposAccion)
            texto: Descripción de la acción realizada
            expediente_id: ID del expediente relacionado (opcional)
            info_adicional: Información adicional en formato dict (se serializa a JSON)
            
        Returns:
            T_Bitacora: Registro de bitácora creado
        """
        try:
            # Serializar info_adicional a JSON si existe
            info_json = None
            if info_adicional:
                info_json = json.dumps(info_adicional, ensure_ascii=False)
            
            # Crear registro usando el repository
            bitacora = self.repo.crear(
                db=db,
                usuario_id=usuario_id,
                tipo_accion_id=tipo_accion_id,
                texto=texto,
                expediente_id=expediente_id,
                info_adicional=info_json
            )
            
            logger.info(
                f"Bitácora registrada: Usuario={usuario_id}, "
                f"Tipo={DESCRIPCIONES_TIPOS_ACCION.get(tipo_accion_id, 'Desconocido')}, "
                f"Texto='{texto}'"
            )
            
            return bitacora
            
        except Exception as e:
            logger.error(f"Error registrando en bitácora: {e}")
            raise Exception(f"Error al registrar en bitácora: {str(e)}")
    
    
    def obtener_con_filtros(
        self,
        db: Session,
        usuario_id: Optional[str] = None,
        tipo_accion_id: Optional[int] = None,
        expediente_numero: Optional[str] = None,
        fecha_inicio: Optional[datetime] = None,
        fecha_fin: Optional[datetime] = None,
        limite: int = 200
    ) -> list:
        """
        Obtiene registros de bitácora con filtros múltiples.
        Compatible con FiltrosBitacora.jsx del frontend.
        
        Args:
            db: Sesión de base de datos
            usuario_id: Filtrar por usuario específico (nombre o correo)
            tipo_accion_id: Filtrar por tipo de acción
            expediente_numero: Filtrar por número de expediente
            fecha_inicio: Fecha de inicio del rango
            fecha_fin: Fecha de fin del rango
            limite: Número máximo de registros
            
        Returns:
            list: Lista de registros de bitácora con información expandida
        """
        try:
            registros = self.repo.obtener_con_filtros(
                db=db,
                usuario_id=usuario_id,
                tipo_accion_id=tipo_accion_id,
                expediente_numero=expediente_numero,
                fecha_inicio=fecha_inicio,
                fecha_fin=fecha_fin,
                limite=limite
            )
            
            # Expandir información para el frontend
            return [self._expandir_registro(r) for r in registros]
            
        except Exception as e:
            logger.error(f"Error obteniendo bitácora con filtros: {e}")
            return []
    
    
    def obtener_estadisticas(self, db: Session, dias: int = 30) -> Dict[str, Any]:
        """
        Obtiene estadísticas agregadas de la bitácora.
        Compatible con DashboardEstadisticas.jsx del frontend.
        
        Args:
            db: Sesión de base de datos
            dias: Número de días hacia atrás para las estadísticas
            
        Returns:
            Dict con estadísticas completas para el dashboard
        """
        try:
            ahora = datetime.utcnow()
            inicio_hoy = ahora.replace(hour=0, minute=0, second=0, microsecond=0)
            inicio_7dias = ahora - timedelta(days=7)
            inicio_30dias = ahora - timedelta(days=30)
            
            # Total de registros
            total_registros = self.repo.contar_total(db)
            
            # Registros por período
            registros_hoy = self.repo.contar_por_periodo(db, inicio_hoy, ahora)
            registros_7dias = self.repo.contar_por_periodo(db, inicio_7dias, ahora)
            registros_30dias = self.repo.contar_por_periodo(db, inicio_30dias, ahora)
            
            # Contadores únicos
            usuarios_unicos = self.repo.contar_usuarios_unicos(db, inicio_30dias, ahora)
            expedientes_unicos = self.repo.contar_expedientes_unicos(db, inicio_30dias, ahora)
            
            # Acciones por tipo (con nombres legibles)
            acciones_por_tipo_raw = self.repo.obtener_acciones_por_tipo(db, inicio_30dias, ahora)
            acciones_por_tipo = [
                {
                    "tipo": DESCRIPCIONES_TIPOS_ACCION.get(item["tipo_id"], f"Tipo {item['tipo_id']}"),
                    "cantidad": item["cantidad"]
                }
                for item in acciones_por_tipo_raw
            ]
            
            return {
                "registrosHoy": registros_hoy,
                "registros7Dias": registros_7dias,
                "registros30Dias": registros_30dias,
                "totalRegistros": total_registros,
                "usuariosUnicos": usuarios_unicos,
                "expedientesUnicos": expedientes_unicos,
                "accionesPorTipo": acciones_por_tipo
            }
            
        except Exception as e:
            logger.error(f"Error obteniendo estadísticas: {e}")
            return {
                "registrosHoy": 0,
                "registros7Dias": 0,
                "registros30Dias": 0,
                "totalRegistros": 0,
                "usuariosUnicos": 0,
                "expedientesUnicos": 0,
                "accionesPorTipo": []
            }
    
    
    def _expandir_registro(self, bitacora: T_Bitacora) -> Dict[str, Any]:
        """
        Expande un registro de bitácora con información de relaciones.
        Compatible con la estructura esperada por el frontend.
        
        Args:
            bitacora: Registro de bitácora
            
        Returns:
            Dict con información expandida
        """
        return {
            "id": bitacora.CN_Id_bitacora,
            "fechaHora": bitacora.CF_Fecha_hora.isoformat() if bitacora.CF_Fecha_hora else None,
            "texto": bitacora.CT_Texto,
            "informacionAdicional": bitacora.CT_Informacion_adicional,
            
            # IDs originales
            "idUsuario": bitacora.CN_Id_usuario,
            "idTipoAccion": bitacora.CN_Id_tipo_accion,
            "idExpediente": bitacora.CN_Id_expediente,
            
            # Información expandida (si existe)
            "usuario": bitacora.usuario.CT_Nombre_usuario if bitacora.usuario else None,
            "correoUsuario": bitacora.usuario.CT_Correo if bitacora.usuario else None,
            "rolUsuario": bitacora.usuario.rol.CT_Nombre_rol if bitacora.usuario and bitacora.usuario.rol else None,
            "tipoAccion": DESCRIPCIONES_TIPOS_ACCION.get(bitacora.CN_Id_tipo_accion) if bitacora.CN_Id_tipo_accion else None,
            "expediente": bitacora.expediente.CT_Num_expediente if bitacora.expediente else None,
            
            # Estado (por defecto Procesado ya que se registró exitosamente)
            "estado": "Procesado"
        }
    
    
    def parsear_info_adicional(self, bitacora: T_Bitacora) -> Optional[Dict[str, Any]]:
        """
        Parsea el campo CT_Informacion_adicional de JSON a dict.
        
        Args:
            bitacora: Registro de bitácora
            
        Returns:
            Optional[Dict]: Información adicional parseada o None
        """
        if not bitacora.CT_Informacion_adicional:
            return None
        
        try:
            return json.loads(bitacora.CT_Informacion_adicional)
        except json.JSONDecodeError as e:
            logger.warning(f"Error parseando info_adicional de bitácora {bitacora.CN_Id_bitacora}: {e}")
            return None


# Instancia singleton del servicio
bitacora_service = BitacoraService()

File: frontend/services/bitacoraService.js
/**
 * bitacoraService.js - Servicio de Bitácora
 * 
 * Maneja operaciones relacionadas con la auditoría y bitácora del sistema:
 * - Consulta de registros con filtros
 * - Estadísticas agregadas
 * - Historial de usuario
 * - Historial por expediente
 */

import httpService from './httpService';

class BitacoraService {
  constructor() {
    this.baseURL = '/bitacora';
  }

  /**
   * Obtener registros de bitácora con filtros múltiples y paginación server-side
   * Solo accesible para administradores
   * 
   * @param {Object} filtros - Filtros de búsqueda
   * @param {string} filtros.usuario - Nombre o correo de usuario
   * @param {number} filtros.tipoAccion - ID del tipo de acción (1-15)
   * @param {string} filtros.expediente - Número de expediente
   * @param {string} filtros.fechaInicio - Fecha inicio ISO
   * @param {string} filtros.fechaFin - Fecha fin ISO
   * @param {number} filtros.page - Número de página (default: 1)
   * @param {number} filtros.limit - Registros por página (default: 10, max: 100)
   * @returns {Promise<Object>} { items: Array, total: number, page: number, pages: number }
   */
  async obtenerRegistros(filtros = {}) {
    try {
      // Construir query params (solo incluir los que tienen valor)
      const params = new URLSearchParams();
      
      if (filtros.usuario) params.append('usuario', filtros.usuario);
      if (filtros.tipoAccion) params.append('tipoAccion', filtros.tipoAccion);
      if (filtros.expediente) params.append('expediente', filtros.expediente);
      
      // Convertir fechas a formato ISO completo (con hora) para FastAPI
      if (filtros.fechaInicio) {
        // Agregar 00:00:00 al inicio del día
        params.append('fechaInicio', `${filtros.fechaInicio}T00:00:00`);
      }
      if (filtros.fechaFin) {
        // Agregar 23:59:59 al final del día
        params.append('fechaFin', `${filtros.fechaFin}T23:59:59`);
      }
      
      // Paginación
      const page = filtros.page || 1;
      const limit = filtros.limit || 10;
      const offset = (page - 1) * limit;
      
      params.append('limite', limit);
      params.append('offset', offset);

      const queryString = params.toString();
      const url = `${this.baseURL}/registros${queryString ? `?${queryString}` : ''}`;

      const data = await httpService.get(url);
    
      // Mapear campos del backend al formato del frontend
      return {
        registros: this._mapearRegistros(data.items || []),
        total: data.total || 0,
        page: data.page || 1,
        pages: data.pages || 1,
        limit: data.limit || limit
      };
      
    } catch (error) {
      console.error('Error obteniendo registros de bitácora:', error);
      
      // Extraer mensaje de error detallado
      let mensajeError = 'Error al obtener registros de bitácora';
      
      if (error.data?.detail) {
        // FastAPI 422 devuelve detail como array o string
        if (Array.isArray(error.data.detail)) {
          mensajeError = error.data.detail.map(e => e.msg || e.message).join(', ');
        } else {
          mensajeError = error.data.detail;
        }
      } else if (error.message) {
        mensajeError = error.message;
      }
      
      throw new Error(mensajeError);
    }
  }

  /**
   * Obtener estadísticas agregadas de la bitácora
   * Solo accesible para administradores
   */
  async obtenerEstadisticas(dias = 30) {
    try {
      const data = await httpService.get(
        `${this.baseURL}/estadisticas?dias=${dias}`
      );
      
      return {
        registrosHoy: data.registrosHoy || 0,
        registros7Dias: data.registros7Dias || 0,
        registros30Dias: data.registros30Dias || 0,
        totalRegistros: data.totalRegistros || 0,
        usuariosUnicos: data.usuariosUnicos || 0,
        expedientesUnicos: data.expedientesUnicos || 0,
        accionesPorTipo: data.accionesPorTipo || [],
        usuariosMasActivos: data.usuariosMasActivos || [],
        expedientesMasConsultados: data.expedientesMasConsultados || [],
        actividadPorDia: data.actividadPorDia || []
      };
      
    } catch (error) {
      console.error('Error obteniendo estadísticas de bitácora:', error);
      throw new Error(
        error.message || 'Error al obtener estadísticas'
      );
    }
  }

  /**
   * Obtener historial del usuario autenticado
   * Accesible para usuarios judiciales y administradores
   */
  async obtenerMiHistorial(limite = 100, tipoAccion = null) {
    try {
      const params = new URLSearchParams({ limite: limite.toString() });
      if (tipoAccion) params.append('tipoAccion', tipoAccion);

      const data = await httpService.get(
        `${this.baseURL}/mi-historial?${params.toString()}`
      );
      
      return this._mapearRegistros(data);
      
    } catch (error) {
      console.error('Error obteniendo mi historial:', error);
      throw new Error(
        error.message || 'Error al obtener tu historial'
      );
    }
  }

  /**
   * Registrar exportación de bitácora
   * Llama al backend para registrar la acción de exportación
   */
  async registrarExportacion() {
    try {
      await httpService.post(`${this.baseURL}/registrar-exportacion`);
    } catch (error) {
      // No lanzamos error para no bloquear la exportación
      console.error('Error registrando exportación en bitácora:', error);
    }
  }

  /**
   * Obtener historial de un expediente específico
   * Muestra todas las acciones realizadas sobre ese expediente
   */
  async obtenerHistorialExpediente(expedienteNumero, limite = 100) {
    try {
      if (!expedienteNumero) {
        throw new Error('Número de expediente es requerido');
      }

      const data = await httpService.get(
        `${this.baseURL}/expediente/${encodeURIComponent(expedienteNumero)}?limite=${limite}`
      );
      
      return this._mapearRegistros(data);
      
    } catch (error) {
      console.error(`Error obteniendo historial del expediente ${expedienteNumero}:`, error);
      throw new Error(
        error.message || 'Error al obtener historial del expediente'
      );
    }
  }

  /**
   * Mapear registros del formato backend al formato frontend
   * Backend usa nomenclatura CN_*, CT_*, CF_*
   * Frontend usa camelCase
   */
  _mapearRegistros(registros) {
    if (!Array.isArray(registros)) return [];

    return registros.map(registro => {
      const mapeado = {
        id: registro.CN_Id_bitacora || registro.id_bitacora || registro.id,
        fechaHora: registro.CF_Fecha_hora || registro.fecha_hora || registro.fechaHora,
        texto: registro.CT_Texto || registro.texto,
        informacionAdicional: this._parseJSON(
          registro.CT_Informacion_adicional || registro.info_adicional || registro.informacionAdicional
        ),
        
        // IDs relacionados
        idUsuario: registro.CN_Id_usuario || registro.id_usuario || registro.idUsuario,
        idTipoAccion: registro.CN_Id_tipo_accion || registro.id_tipo_accion || registro.idTipoAccion,
        idExpediente: registro.CN_Id_expediente || registro.id_expediente || registro.idExpediente,
        
        // Datos relacionados (si están expandidos)
        usuario: registro.usuario || registro.nombre_usuario || 'Usuario desconocido',
        correoUsuario: registro.correoUsuario || registro.correo_usuario || null,
        tipoAccion: registro.tipoAccion || registro.tipo_accion || registro.nombre_tipo_accion || this._mapearTipoAccion(registro.idTipoAccion || registro.CN_Id_tipo_accion || registro.id_tipo_accion),
        expediente: registro.expediente || registro.numero_expediente || null,
        
        // Campos adicionales útiles para UI
        rolUsuario: registro.rol_usuario || registro.rolUsuario || null,
        estado: registro.estado || 'Procesado',
        ip: registro.ip || null,
        navegador: registro.navegador || null,
        duracion: registro.duracion || null
      };

      return mapeado;
    });
  }

  /**
   * Parsear JSON de forma segura
   */
  _parseJSON(jsonString) {
    if (!jsonString) return null;
    if (typeof jsonString === 'object') return jsonString;
    
    try {
      return JSON.parse(jsonString);
    } catch (e) {
      return jsonString;
    }
  }

  /**
   * Mapear ID de tipo de acción a nombre legible
   */
  _mapearTipoAccion(idTipo) {
    const tipos = {
      1: 'Búsqueda de Casos Similares',
      2: 'Carga de Documentos',
      3: 'Login',
      4: 'Logout',
      5: 'Cambio de Contraseña',
      6: 'Recuperación de Contraseña',
      7: 'Crear Usuario',
      8: 'Editar Usuario',
      9: 'Consultar Usuarios',
      10: 'Descargar Archivo',
      11: 'Listar Archivos',
      12: 'Consulta RAG',
      13: 'Generar Resumen',
      14: 'Consultar Bitácora',
      15: 'Exportar Bitácora'
    };
    
    return tipos[idTipo] || `Acción ${idTipo}`;
  }

  /**
   * Exportar registros a CSV (funcionalidad futura)
   */
  async exportarRegistros(filtros = {}) {
    // TODO: Implementar cuando el backend tenga el endpoint de exportación
    throw new Error('Funcionalidad de exportación pendiente de implementación');
  }
}

// Exportar instancia singleton
const bitacoraService = new BitacoraService();
export default bitacoraService;

// Exportar métodos individuales para imports selectivos
export const obtenerRegistros = (filtros) => bitacoraService.obtenerRegistros(filtros);
export const obtenerEstadisticas = (dias) => bitacoraService.obtenerEstadisticas(dias);
export const obtenerMiHistorial = (limite, tipoAccion) => bitacoraService.obtenerMiHistorial(limite, tipoAccion);
export const obtenerHistorialExpediente = (expediente, limite) => bitacoraService.obtenerHistorialExpediente(expediente, limite);
